opal/mca/btl/ofi/btl_ofi_module.c

```diff
@@ -0,0 +1,296 @@
+    mca_btl_ofi_reg_t *reg;
+    int access_flags = flags & MCA_BTL_REG_FLAG_ACCESS_ANY;
+    int rc;
+
+    rc = ofi_module->rcache->rcache_register (ofi_module->rcache, base, size, 0, access_flags,
+                                              (mca_rcache_base_registration_t **) &reg);
+    if (OPAL_UNLIKELY(OPAL_SUCCESS != rc)) {
+        return NULL;
+    }
+
+    return &reg->handle;
+}
+
+/**
+ * @brief Deregister a memory region
+ *
+ * @param btl (IN)         BTL module region was registered with
+ * @param handle (IN)      BTL registration handle to deregister
+ *
+ * This function deregisters the memory region associated with the specified handle. Care
+ * should be taken to not perform any RDMA or atomic operation on this memory region
+ * after it is deregistered. It is erroneous to specify a memory handle associated with
+ * a remote node.
+ */
+static int mca_btl_ofi_deregister_mem (mca_btl_base_module_t *btl, mca_btl_base_registration_handle_t *handle)
+{
+    mca_btl_ofi_module_t *ofi_module = (mca_btl_ofi_module_t *) btl;
+    mca_btl_ofi_reg_t *reg =
+        (mca_btl_ofi_reg_t *)((intptr_t) handle - offsetof (mca_btl_ofi_reg_t, handle));
+
+    (void) ofi_module->rcache->rcache_deregister (ofi_module->rcache, &reg->base);
+
+    return OPAL_SUCCESS;
+}
+
+int mca_btl_ofi_reg_mem (void *reg_data, void *base, size_t size, mca_rcache_base_registration_t *reg)
+{
+    int rc;
+    static uint64_t access_flags = FI_REMOTE_WRITE | FI_REMOTE_READ | FI_READ | FI_WRITE;
+
+    mca_btl_ofi_module_t *btl = (mca_btl_ofi_module_t*) reg_data;
+    mca_btl_ofi_reg_t *ur = (mca_btl_ofi_reg_t*) reg;
+
+    rc = fi_mr_reg(btl->domain, base, size, access_flags, 0,
+                   (uint64_t) reg, 0, &ur->ur_mr, NULL);
+    if (0 != rc) {
+        return OPAL_ERR_OUT_OF_RESOURCE;
+    }
+
+    ur->handle.rkey = fi_mr_key(ur->ur_mr);
+    ur->handle.desc = fi_mr_desc(ur->ur_mr);
+
+    /* In case the provider doesn't support FI_MR_VIRT_ADDR,
+     * we need to reference the remote address by the distance from base registered
+     * address. We keep this information to use in rdma/atomic operations. */
+    if (btl->use_virt_addr) {
+        ur->handle.base_addr = 0;
+    } else {
+        ur->handle.base_addr = base;
+    }
+
+    return OPAL_SUCCESS;
+}
+
+int mca_btl_ofi_dereg_mem (void *reg_data, mca_rcache_base_registration_t *reg)
+{
+    mca_btl_ofi_reg_t *ur = (mca_btl_ofi_reg_t*)reg;
+
+    if (ur->ur_mr != NULL) {
+        if (0 != fi_close(&ur->ur_mr->fid)) {
+            opal_output(0, "%s: error unpinning memory mr=%p: %s",
+                    __func__, (void*) ur->ur_mr, strerror(errno));
+            return OPAL_ERROR;
+        }
+    }
+
+    return OPAL_SUCCESS;
+}
+
+/*
+ * Cleanup/release module resources.
+ */
+
+int mca_btl_ofi_finalize (mca_btl_base_module_t* btl)
+{
+    mca_btl_ofi_module_t *ofi_btl = (mca_btl_ofi_module_t *) btl;
+    mca_btl_ofi_endpoint_t *endpoint, *next;
+
+    fi_close(&ofi_btl->cq->fid);
```

Comment:
These close and freeinfo statements might want to have conditionals around them to ensure that various pointers aren't NULL.  E.g.,

```c
if (NULL != ofi_btl->cq) {
    fi_close(&obi_btl->cq->fid);
}
```

just in case finalize is called early and somehow these pointers haven't been setup yet.  I.e., make it safe to call finalize at any time / at any point during the setup.

Reply:
fixed.